File: apps/api/src/routes/catalog.ts
import Router from "@koa/router";
import { z } from "zod";
import type { AppDependencies } from "../container.js";
import { createAuthMiddleware, requireAuthUser } from "../middleware/auth.js";
import { getCriticScore } from "../services/criticScoreData.js";

const listQuerySchema = z.object({
  q: z.string().optional(),
  cursor: z.string().optional(),
  limit: z.coerce.number().int().min(1).max(100).default(12)
});

const detailParamsSchema = z.object({
  externalId: z.string().min(1)
});

const gameRatingSchema = z.object({
  score: z.coerce.number().min(0.1).max(10).refine((value) => Number.isInteger(value * 10), {
    message: "Score must use 0.1 increments"
  })
});

function isInternalRequest(ctx: Router.RouterContext, internalToken: string): boolean {
  const token = ctx.headers["x-internal-token"];
  return typeof token === "string" && token === internalToken;
}

export function createCatalogRouter(deps: AppDependencies): Router {
  const router = new Router();
  const requireAuth = createAuthMiddleware(deps.env);

  router.get("/api/catalog/games", requireAuth, async (ctx) => {
    const parsed = listQuerySchema.safeParse(ctx.query);
    if (!parsed.success) {
      ctx.status = 400;
      ctx.body = { message: "Invalid query", issues: parsed.error.flatten() };
      return;
    }

    const authUser = requireAuthUser(ctx.state);
    const [catalog, ownedGames] = await Promise.all([
      deps.catalogService.listCatalog({
        query: parsed.data.q,
        cursor: parsed.data.cursor,
        limit: parsed.data.limit
      }),
      deps.repository.listGamesByUserId(authUser.userId)
    ]);

    const ownedByExternalId = new Map(
      ownedGames.map((game) => [game.externalId, { gameId: game.id, ownedAt: game.ownedAt }])
    );

    ctx.body = {
      items: catalog.items.map((item) => {
        const owned = ownedByExternalId.get(item.externalId);
        return {
          ...item,
          isOwned: Boolean(owned),
          ownedGameId: owned?.gameId ?? null,
          ownedAt: owned?.ownedAt ?? null
        };
      }),
      nextCursor: catalog.nextCursor,
      totalCount: catalog.totalCount
    };
  });

  router.get("/api/catalog/games/:externalId", requireAuth, async (ctx) => {
    const parsed = detailParamsSchema.safeParse(ctx.params);
    if (!parsed.success) {
      ctx.status = 400;
      ctx.body = { message: "Invalid params", issues: parsed.error.flatten() };
      return;
    }

    const authUser = requireAuthUser(ctx.state);
    const [catalogGame, ownedGames, playerRating] = await Promise.all([
      deps.catalogService.getCatalogGame(parsed.data.externalId),
      deps.repository.listGamesByUserId(authUser.userId),
      deps.playtimeService.getPlayerRatingByExternalId(authUser.userId, parsed.data.externalId)
    ]);

    if (!catalogGame) {
      ctx.status = 404;
      ctx.body = { message: "Catalog game not found" };
      return;
    }

    const ownedGame = ownedGames.find((entry) => entry.externalId === parsed.data.externalId) ?? null;
    const [effectivePlaytime, corrections] = ownedGame
      ? await Promise.all([
          deps.playtimeService.getEffectiveByGameId(authUser.userId, ownedGame.id),
          deps.playtimeService.listCorrections(authUser.userId, ownedGame.id)
        ])
      : [null, []];

    ctx.body = {
      ...catalogGame,
      criticScore: getCriticScore(parsed.data.externalId),
      playerRating,
      ownedGame: ownedGame
        ? {
            id: ownedGame.id,
            externalId: ownedGame.externalId,
            title: ownedGame.title,
            coverUrl: ownedGame.coverUrl,
            ownedAt: ownedGame.ownedAt,
            lastPlayedAt: ownedGame.lastPlayedAt,
            priceAmount: ownedGame.priceJpy,
            priceCurrency: "USD",
            effectivePlaytime,
            localizations: catalogGame.localizations
          }
        : null,
      corrections
    };
  });

  router.put("/api/catalog/games/:externalId/rating", requireAuth, async (ctx) => {
    const parsedParams = detailParamsSchema.safeParse(ctx.params);
    const parsedBody = gameRatingSchema.safeParse(ctx.request.body);
    if (!parsedParams.success || !parsedBody.success) {
      ctx.status = 400;
      ctx.body = {
        message: "Invalid payload",
        issues: {
          params: parsedParams.success ? undefined : parsedParams.error.flatten(),
          body: parsedBody.success ? undefined : parsedBody.error.flatten()
        }
      };
      return;
    }

    const authUser = requireAuthUser(ctx.state);
    const catalogGame = await deps.catalogService.getCatalogGame(parsedParams.data.externalId);
    if (!catalogGame) {
      ctx.status = 404;
      ctx.body = { message: "Catalog game not found" };
      return;
    }

    const rating = await deps.playtimeService.rateGameByExternalId({
      userId: authUser.userId,
      externalId: parsedParams.data.externalId,
      score: parsedBody.data.score
    });
    ctx.body = { rating };
  });

  router.get("/api/catalog/status", requireAuth, async (ctx) => {
    const status = await deps.catalogService.getCatalogStatus();
    ctx.body = { status };
  });

  router.post("/api/internal/catalog/refresh", async (ctx) => {
    if (!isInternalRequest(ctx, deps.env.INTERNAL_SYNC_TOKEN)) {
      ctx.status = 401;
      ctx.body = { message: "Unauthorized internal call" };
      return;
    }

    const result = await deps.catalogService.refreshCatalog();
    ctx.body = { result };
  });

  return router;
}

File: backend/internal/handler/catalog.go
package handler

import (
	"encoding/base64"
	"encoding/json"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"nintendo-gametime/internal/domain"
	"nintendo-gametime/internal/middleware"
	"nintendo-gametime/internal/repository"
	"nintendo-gametime/internal/rvis"
)

type CatalogHandler struct {
	repo repository.Repository
	rvis *rvis.Service
}

func NewCatalogHandler(repo repository.Repository, rvisSvc *rvis.Service) *CatalogHandler {
	return &CatalogHandler{repo: repo, rvis: rvisSvc}
}

func (h *CatalogHandler) ListCatalog(c *gin.Context) {
	auth := middleware.GetAuthUser(c)
	query := c.Query("q")
	limit := clampInt(queryInt(c, "limit", 20), 1, 100)
	cursor := c.Query("cursor")

	offset := 0
	if cursor != "" {
		if decoded, err := base64.RawURLEncoding.DecodeString(cursor); err == nil {
			offset, _ = strconv.Atoi(string(decoded))
		}
	}

	games, err := h.repo.ListCatalogGames(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to list catalog"})
		return
	}

	// Get user's owned games for isOwned check
	userGames, _ := h.repo.ListGamesByUserID(c.Request.Context(), auth.UserID)
	ownedMap := make(map[string]string, len(userGames))
	for _, ug := range userGames {
		ownedMap[ug.ExternalID] = ug.ID
	}

	// Filter by query if provided
	var filtered []domain.CatalogGameRow
	if query != "" {
		for _, g := range games {
			if containsIgnoreCase(g.Title, query) {
				filtered = append(filtered, g)
			}
		}
	} else {
		filtered = games
	}

	totalCount := len(filtered)

	// Apply pagination
	if offset >= len(filtered) {
		c.JSON(http.StatusOK, gin.H{"items": []gin.H{}, "nextCursor": nil, "totalCount": totalCount})
		return
	}
	end := offset + limit
	if end > len(filtered) {
		end = len(filtered)
	}
	page := filtered[offset:end]

	// Build CatalogItem response
	items := make([]gin.H, 0, len(page))
	for _, g := range page {
		ownedGameID := ownedMap[g.ExternalID]
		isOwned := ownedGameID != ""

		// Parse localizations
		var localizations gin.H
		if g.Localizations != nil {
			var loc map[string]interface{}
			if err := json.Unmarshal(g.Localizations, &loc); err == nil {
				localizations = gin.H(loc)
			}
		}

		item := gin.H{
			"externalId":    g.ExternalID,
			"title":         g.Title,
			"coverUrl":      g.CoverURL,
			"storeUrl":      g.StoreURL,
			"description":   g.Description,
			"publisher":     g.Publisher,
			"releaseDate":   g.ReleaseDate,
			"priceAmount":   g.PriceAmount,
			"priceCurrency": g.PriceCurrency,
			"platform":      g.Platform,
			"region":        g.Region,
			"localizations": localizations,
			"isOwned":       isOwned,
			"ownedGameId":   ownedGameID,
		}
		items = append(items, item)
	}

	var nextCursor *string
	if end < len(filtered) {
		s := base64.RawURLEncoding.EncodeToString([]byte(strconv.Itoa(end)))
		nextCursor = &s
	}

	c.JSON(http.StatusOK, gin.H{"items": items, "nextCursor": nextCursor, "totalCount": totalCount})
}

func (h *CatalogHandler) GetCatalogGame(c *gin.Context) {
	auth := middleware.GetAuthUser(c)
	ctx := c.Request.Context()
	externalID := c.Param("externalId")

	game, err := h.repo.GetCatalogGameByExternalID(ctx, externalID)
	if err != nil || game == nil {
		c.JSON(http.StatusNotFound, gin.H{"message": "Catalog game not found"})
		return
	}

	// Parse localizations
	var localizations gin.H
	if game.Localizations != nil {
		var loc map[string]interface{}
		if err := json.Unmarshal(game.Localizations, &loc); err == nil {
			localizations = gin.H(loc)
		}
	}

	// Check if user owns this game
	var ownedGame *gin.H
	userGames, _ := h.repo.ListGamesByUserID(ctx, auth.UserID)
	for _, ug := range userGames {
		if ug.ExternalID == game.ExternalID {
			// Build enriched owned game
			snapshots, _ := h.repo.GetLatestOfficialSnapshotsByUserID(ctx, auth.UserID)
			corrections, _ := h.repo.ListCorrectionsByUserID(ctx, auth.UserID, nil)
			playtimeMap := domain.CalculateEffectivePlaytimeMap(snapshots, corrections)
			pt := playtimeMap[ug.ID]
			if pt.TotalMinutes == 0 && pt.Source == "" {
				pt = domain.EffectivePlaytime{GameID: ug.ID, Source: "official"}
			}
			og := gin.H{
				"id":        ug.ID,
				"externalId": ug.ExternalID,
				"title":     ug.Title,
				"coverUrl":  ug.CoverURL,
				"ownedAt":   ug.OwnedAt,
				"lastPlayedAt": ug.LastPlayedAt,
				"platform":      game.Platform,
				"region":        game.Region,
				"localizations": localizations,
				"priceAmount":  ug.PriceJPY,
				"priceCurrency": "JPY",
				"effectivePlaytime": gin.H{
					"totalMinutes":          pt.TotalMinutes,
					"officialMinutes":       pt.OfficialMinutes,
					"correctionDeltaMinutes": pt.CorrectionDeltaMinutes,
					"source":                pt.Source,
					"updatedAt":             pt.UpdatedAt,
				},
			}
			ownedGame = &og
			break
		}
	}

	// Player rating
	ur, sum, _ := h.repo.GetGameRatingSnapshot(ctx, auth.UserID, game.ExternalID)
	playerRating := gin.H{"userScore": 0, "averageScore": 0, "ratingCount": 0}
	if ur != nil {
		playerRating["userScore"] = ur.Score
	}
	if sum != nil && sum.RatingCount > 0 {
		playerRating["averageScore"] = sum.RatingTotal / float64(sum.RatingCount)
		playerRating["ratingCount"] = sum.RatingCount
	}

	result := gin.H{
		"externalId":    game.ExternalID,
		"title":         game.Title,
		"coverUrl":      game.CoverURL,
		"storeUrl":      game.StoreURL,
		"description":   game.Description,
		"publisher":     game.Publisher,
		"releaseDate":   game.ReleaseDate,
		"priceAmount":   game.PriceAmount,
		"priceCurrency": game.PriceCurrency,
		"platform":      game.Platform,
		"region":        game.Region,
		"localizations": localizations,
		"criticScore":   game.CriticScore,
		"playerRating":  playerRating,
		"ownedGame":     ownedGame,
		"corrections":   []gin.H{},
	}

	c.JSON(http.StatusOK, result)
}

func (h *CatalogHandler) RateCatalogGame(c *gin.Context) {
	auth := middleware.GetAuthUser(c)
	externalID := c.Param("externalId")
	var req struct {
		Score float64 `json:"score" binding:"required,min=0.1,max=10"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid payload"})
		return
	}

	ur, sum, err := h.repo.UpsertGameRating(c.Request.Context(), auth.UserID, externalID, req.Score, time.Now().UTC().Format(time.RFC3339))
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to rate game"})
		return
	}

	rating := gin.H{"userScore": 0, "averageScore": 0, "ratingCount": 0}
	if ur != nil {
		rating["userScore"] = ur.Score
	}
	if sum != nil && sum.RatingCount > 0 {
		rating["averageScore"] = sum.RatingTotal / float64(sum.RatingCount)
		rating["ratingCount"] = sum.RatingCount
	}
	c.JSON(http.StatusOK, gin.H{"rating": rating})
}

func (h *CatalogHandler) GetCatalogStatus(c *gin.Context) {
	count, err := h.repo.CountCatalogGames(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to get status"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"totalGames": count, "lastSynced": time.Now().UTC().Format(time.RFC3339)})
}

func (h *CatalogHandler) GetPrices(c *gin.Context) {
	externalID := c.Param("externalId")
	prices, err := h.repo.GetRegionalPrices(c.Request.Context(), externalID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to get prices"})
		return
	}
	if prices == nil {
		c.JSON(http.StatusOK, gin.H{"prices": []interface{}{}})
		return
	}
	c.JSON(http.StatusOK, prices)
}

func (h *CatalogHandler) GetPriceMap(c *gin.Context) {
	externalID := c.Param("externalId")
	priceRow, err := h.repo.GetRegionalPrices(c.Request.Context(), externalID)
	if err != nil || priceRow == nil {
		c.JSON(http.StatusNotFound, gin.H{"message": "No price data found"})
		return
	}

	// Get game title for the map
	game, _ := h.repo.GetCatalogGameByExternalID(c.Request.Context(), externalID)
	title := "eShop Price Map"
	if game != nil && game.Title != "" {
		title = game.Title
	}

	var prices []domain.RegionalPrice
	if err := json.Unmarshal(priceRow.Prices, &prices); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to parse price data"})
		return
	}

	html, err := h.rvis.RenderPriceMap(prices, title)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Failed to render price map", "error": err.Error()})
		return
	}

	c.Data(http.StatusOK, "text/html; charset=utf-8", []byte(html))
}

func containsIgnoreCase(s, sub string) bool {
	return len(s) >= len(sub) && (s == sub || len(sub) == 0 ||
		(len(s) > 0 && len(sub) > 0 && containsLower(s, sub)))
}

func containsLower(s, sub string) bool {
	for i := 0; i <= len(s)-len(sub); i++ {
		match := true
		for j := 0; j < len(sub); j++ {
			sc := s[i+j]
			tc := sub[j]
			if sc >= 'A' && sc <= 'Z' {
				sc += 32
			}
			if tc >= 'A' && tc <= 'Z' {
				tc += 32
			}
			if sc != tc {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}
